File: 백준/Gold/9019. DSLR/DSLR.java
import java.io.*;
import java.util.*;

public class Main {
    static class Register {
        int num;
        String command;

        public Register(int num, String command) {
            this.num = num;
            this.command = command;
        }

        int d() {
            return (num * 2) % 10000;
        }

        int s() {
            return num == 0 ? 9999 : num - 1;
        }

        int l() {
            return num % 1000 * 10 + num / 1000;
        }

        int r() {
            return num % 10 * 1000 + num / 10;
        }
    }
    public static String bfs(int start, int target) {
        boolean[] visit = new boolean[10000];

        Queue<Register> q = new LinkedList<>();
        q.offer(new Register(start, ""));
        visit[start] = true;

        while (!q.isEmpty()) {
            Register now = q.poll();
            if (target == now.num) {
                return now.command;
            }

            if (!visit[now.d()]) {
                q.add(new Register(now.d(), now.command + "D"));
                visit[now.d()] = true;
            }

            if (!visit[now.s()]) {
                q.add(new Register(now.s(), now.command + "S"));
                visit[now.s()] = true;
            }

            if (!visit[now.l()]) {
                q.add(new Register(now.l(), now.command + "L"));
                visit[now.l()] = true;
            }

            if (!visit[now.r()]) {
                q.add(new Register(now.r(), now.command + "R"));
                visit[now.r()] = true;
            }
        }

        return "";
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder stb = new StringBuilder();
        int t = Integer.parseInt(br.readLine());

        while (t-- > 0) {
            String[] input = br.readLine().split(" ");
            stb.append(bfs(Integer.parseInt(input[0]), Integer.parseInt(input[1]))).append("\n");
        }

        System.out.print(stb.toString());
    }
}

File: 백준/Gold/9019. DSLR/DSLR.js
let [t, ...input] = require("fs")
  .readFileSync(process.platform === "linux" ? "/dev/stdin" : "input.txt")
  .toString()
  .trim()
  .split("\n");

let ans = [];

const bfs = (start, target) => {
  let visit = Array(10000).fill(false);
  let q = [];
  q.push([start, ""]);
  visit[start] = true;

  while (q.length !== 0) {
    let [num, commend] = q.shift();
    if (target === num) {
      return commend;
    }

    let d = (num * 2) % 10000;
    if (!visit[d]) {
      q.push([d, commend + "D"]);
      visit[d] = true;
    }

    let s = num === 0 ? 9999 : num - 1;
    if (!visit[s]) {
      q.push([s, commend + "S"]);
      visit[s] = true;
    }

    let l = (num % 1000) * 10 + Math.floor(num / 1000);
    if (!visit[l]) {
      q.push([l, commend + "L"]);
      visit[l] = true;
    }

    let r = (num % 10) * 1000 + Math.floor(num / 10);
    if (!visit[r]) {
      q.push([r, commend + "R"]);
      visit[r] = true;
    }
  }
};

while (t-- > 0) {
  let [start, target] = input.shift().split(" ");
  ans.push(bfs(Number(start), Number(target)));
}

console.log(ans.join("\n"));
